### PR TITLE
03 data structure

### PR DESCRIPTION
**Pull request by sefrem**

### CODE REVIEW

File: homeworks/03-data-structures/src/BinarySearchTree.test.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import {BinarySearchTree} from "./BinarySearchTree";
+import {testTree} from "./testData";
+
+describe('BinarySearchTree', () => {
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/BinarySearchTree.test.ts:4]: I would expect tests of inherited methods as well.

File: homeworks/03-data-structures/src/BinarySearchTree.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import {BinaryTree, IBinaryTree} from "./BinaryTree";
+import {TreeNode} from "./types";
+
+interface IBinarySearchTree extends IBinaryTree<number> {
+    has(value: number): boolean;
+}
+
+export class BinarySearchTree extends BinaryTree<number> implements IBinarySearchTree {
+
+    has = (value: number): boolean => {
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/BinarySearchTree.ts:10]: Always define access modifiers.

File: homeworks/03-data-structures/src/BinarySearchTree.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        const traverse = (node: TreeNode<number>): TreeNode<number> => {
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/BinarySearchTree.ts:11]: You should keep it in a separate private method.

File: homeworks/03-data-structures/src/BinaryTree.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import {TraverseType, TreeNode} from "./types";
+
+export interface IBinaryTree<T> {
+
+    setTree(value: TreeNode<T>): void,
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/BinaryTree.ts:5]: Doesn't satisfy: `setTree(tree: TreeNode): this;`

File: homeworks/03-data-structures/src/BinaryTree.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+export class BinaryTree<T> implements IBinaryTree<T> {
+
+    constructor(protected tree: TreeNode<T>) {}
+
+    setTree = (value: TreeNode<T>): void => {
+        this.tree = value
+    }
+
+    traverse = (value: TraverseType): T[] => {
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/BinaryTree.ts:20]: Always define access modifiers.

File: homeworks/03-data-structures/src/mergeSort.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        return [...arr, ...left, ...right]
+    }
+
+    function divide(array: T[]): T[] {
+        if (array.length < 2) {
+            return array
+        }
+        const leftHalf = array.splice(0, array.length / 2)
+        return merge(divide(leftHalf), divide(array))
+    }
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/mergeSort.ts:37]: You can keep it outside the function.

File: homeworks/03-data-structures/src/types.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+    value: T,
+    left?: TreeNode<T>,
+    right?: TreeNode<T>,
+}
+
+export enum TraverseType {
+    INORDER = 'inorder',
+    PREORDER = 'preorder',
+    POSTORDER = 'postorder',
+    BFS = 'bfs',
```
Comment (shpakkdv) [inline homeworks/03-data-structures/src/types.ts:11]: Enum naming convention:
image

**Review by shpakkdv — APPROVED** [review]
Looks good 👍

File: homeworks/03-data-structures/src/BinarySearchTree.test.ts (hunk 1 above)
Comment (sefrem) [reply homeworks/03-data-structures/src/BinarySearchTree.test.ts:4]: is it necessary if the test coverage is 100%?

**Pull request merged**